### PR TITLE
std: Stablize the macros module

### PR DESCRIPTION
**Pull request by alexcrichton**
This commit performs a pass over the `std::macros` module, applying stability
attributes where necessary. In particular, this audits macros for patterns such
as:
- Standard use of forward-to-format-args via `$($arg:tt)*` (or `+`)
- Prevent macro-defined identifiers from leaking into expression arguments as
  hygiene is not perfectly implemented.
- Wherever possible, `$crate` is used now.

Specifically, the following actions were taken:
- The `std::macros` module itself is no longer public.
- The `panic!` macro is stable
- The `assert!` macro is stable
- The `assert_eq!` macro is stable
- The `debug_assert!` macro is stable
- The `debug_assert_eq!` macro is stable
- The `unreachable!` macro is stable after removing the extra forms to bring the
  definition in line with the `unimplemented!` macro.
- The `try!` macro is stable
- The `vec!` macro is stable

### CODE REVIEW

File: src/libstd/macros.rs [hunk 1]
```diff
@@ -210,21 +217,15 @@ macro_rules! debug_assert_eq {
...
-    () => ({
-        panic!("internal error: entered unreachable code")
-    });
-    ($msg:expr) => ({
-        unreachable!("{}", $msg)
-    });
-    ($fmt:expr, $($arg:tt)*) => ({
-        panic!(concat!("internal error: entered unreachable code: ", $fmt), $($arg)*)
-    });
+    () => (panic!("internal error: entered unreachable code"))
```
Comment (aturon) [inline src/libstd/macros.rs]: What's the rationale for dropping the functionality here, rather than beefing up `unimplemented`?

**Comment by aturon** [comment]
This basically looks good to me, but we may want to get input from someone with deeper macros knowledge.

Also, are we 100% sure we want to stabilize `vec!`? It's probably so widely-used that we can't ever get rid of it...

File: src/libstd/macros.rs (hunk 1 above)
Comment (alexcrichton) [inline src/libstd/macros.rs]: I'd personally like to remove both of them as they're just longer versions of `panic!` in my mind, but failing that I'd rather stick to a more conservative approach where we don't just duplicate-panic-in-each-macro just yet.

**Comment by alexcrichton** [comment]
With regard to `vec!`, you're thinking of perhaps a more generic collections-creating macro? The current implementation is pretty super-specialized to `Vec` (deconstructing `Box<[T]>` and reassembling), but it is unfortunate that we don't have macros for other containers. Furthermore, it'd be unfortunate to have either:

* A generic macro for all containers, plus a more efficient `vec!` macro
* A macro-per-container

I do agree though that we've put ourselves in a corner with `vec!` right now, we don't really have a great way out of it :(

cc @kmcallister, @sfackler

File: src/libsyntax/ext/base.rs [hunk 2]
```diff
@@ -353,9 +350,6 @@ fn initial_syntax_expander_table(ecfg: &expand::ExpansionConfig) -> SyntaxEnv {
     syntax_expanders.insert(intern("option_env"),
                             builtin_normal_expander(
                                     ext::env::expand_option_env));
-    syntax_expanders.insert(intern("bytes"),
```
Comment (sfackler) [inline src/libsyntax/ext/base.rs]: If this is being removed, shouldn't the ext::bytes module be removed as well?

**Comment by sfackler** [comment]
The commit should be tagged as a `[breaking-change]` since `fmt!`, `include_bin!` and `bytes!` are going away.

**Comment by aturon** [comment]
@alexcrichton Yep, those were pretty much exactly my thoughts. The only question is whether we want to fully commit now, or wait till beta to admit that we have to.

File: src/libsyntax/ext/base.rs (hunk 2 above)
Comment (alexcrichton) [inline src/libsyntax/ext/base.rs]: Aha, indeed!

**Comment by kmcallister** [comment]
I think the best case outcome for `vec!` is a generic `seq!` macro that uses traits to work for multiple kinds of containers, at which point `vec!` could become a synonym.

**Comment by kmcallister** [comment]
These changes look good to me.

**Comment by alexcrichton** [comment]
rebased, re-r? @aturon

**Pull request merged**